### ISSUE TITLE
[clang-tidy] New check for static_cast of an enum class type

### ISSUE THREAD

**Issue by FalcoGer** [llvm/llvm-project#71543]
When casting an `enum class` type using `static_cast<T>`, this may result in a bug if the type `T` is not correct. In c++23 we have `std::to_underlying` which should be used instead. This also is more desirable because if the enum's underlying type is changed later, the casting code is automatically still correct without the need to make the change there, too.

```c++
// Fix, must add, if not already present:
#include <utility> // Provides std::to_underlying()


// NOLINTNEXTLINE(performance-enum-size)
enum class EType : int {};

auto main() -> int
{
    // NOLINTNEXTLINE(readability-magic-numbers)
    EType e {42};

    [[maybe_unused]]
    // c++23 and later only

    // Original code:
    // auto i = static_cast<int>(e);
    //          ~~~~~~~~~~~~~~~~
    //          ^ bugprone: use std::to_underlying
    // Fix:
    auto i = std::to_underlying(e);
    return 0;
}
```

**Comment by carlosgalvezp** [comment]
`modernize-use-std-to-underlying`?

It would be good to have an option for the user to specify their own hand-rolled version of this function, since nobody is going to wait for C++23 for it :) The check should probably not warn about the `static_cast` done inside that function.

**Comment by vbvictor** [comment]
> The check should probably not warn about the static_cast done inside that function.

I think one-time suppression will be enough for this. This function is likely to come from system headers so it will be filtered anyway most of the time.

**Comment by vbvictor** [comment]
Otherwise, we would need to look for `functionDecl` that 99.99% will be useless

**Comment by vbvictor** [comment]
Please read https://clang.llvm.org/extra/clang-tidy/Contributing.html for guides on how to write clang-tidy check.
Most important steps are:

1. Use `add_new_check.py` to booststrap, take name `modernize-use-to-underlying` (I omitted `std` as opposed to @carlosgalvezp suggestion because most of the check don't do that).
2. Create options for custom `to_underlying` function and include header, look https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-std-format.html for reference (`ReplacementFormatFunction` and `FormatHeader` options).
3. Write comprehensive test suite that check both default behavior and custom functions behavior
4. Follow https://clang.llvm.org/extra/clang-tidy/Contributing.html#making-your-check-robust and https://clang.llvm.org/extra/clang-tidy/Contributing.html#making-your-check-robust guides

As an inspiration, there is similar rule in SonarQube https://rules.sonarsource.com/cpp/tag/since-c++23/RSPEC-7035/, but our version will be better and support e.g. `llvm::to_underlying` too :)

**Comment by LegalizeAdulthood** [comment]
> > write `+e`
> 
> Arguably that's a detriment in readability :) I believe we have a check warning about that (for ints only).

Readability is -- as always -- subjective.  I don't know what you consider more readable, but `static_cast<int>(e)` is not what I consider more readable.

Warning about `+e` should never happen when `e` is a type that has a user-defined `operator+`.

**Comment by arjunkonidala** [comment]
my team would like to work on this issue

**Comment by vbvictor** [comment]
@arjunkonidala, Go ahead, I've assigned issue to you.

**Comment by vogelsgesang** [comment]
New attempt in https://github.com/llvm/llvm-project/pull/210459
(since I guess the old PR #167212 is dead and I would also like to use this check)

**Comment by sangam151** [comment]
is there anyone else working on this issue? iam a new contributor can i work on this issue?

**Comment by vbvictor** [comment]
> is there anyone else working on this issue? iam a new contributor can i work on this issue?

Yes, there is PR in review https://github.com/llvm/llvm-project/pull/210459

**Comment by vbvictor** [comment]
Done in https://github.com/llvm/llvm-project/pull/210459
